### ISSUE TITLE
Tool calls silently degraded: truncated tool names, dropped arguments, and no-op Bash substitutions

### ISSUE THREAD

**Issue by fusengine** [MoonshotAI/kimi-code#3220]
# Tool calls silently degraded: truncated tool names, dropped arguments, and no-op `Bash` substitutions

## Environment

| | |
|---|---|
| `kimi` CLI | **0.38.0** |
| Model | `kimi-code/k3` (provider `managed:kimi-code`, `https://api.kimi.com/coding/v1`) |
| Thinking | `enabled = true`, `effort = "high"` (model declares `always_thinking`) |
| OS | macOS 15 (Darwin 25.6.0), arm64 |
| Node | 26.7.0 |
| MCP servers | ~12 connected, ~155 tools in the catalog (~248 KB of schemas per request) |

Also reproduced through the embedded server (`kap-server` + `agent-core-v2`) in a desktop host, so this is not TUI-specific.

## Summary

Two distinct defects make agent runs stall. The first is an engine bug with a proposed patch. The second is a model-side emission problem that the engine currently accepts silently instead of retrying.

---

### Symptom

```
Tool "mcp__fuse-browser__browser_ope" not found
```

The call is recorded with an empty argument object:

```json
{"name":"mcp__fuse-browser__browser_ope","args":{}}
```

The tool catalog sent in the request contains the correct `mcp__fuse-browser__browser_open`. The engine answers `Tool not found`, the model retries the same call, and the turn loops until the step budget is exhausted. Observed 8–16 occurrences per affected session.

### Root cause

`packages/agent-core-v2/src/kosong/provider/bases/openai/chat-completions-stream.ts`, `convertChatCompletionStreamToolCall`.

The buffered branch emits the `ToolCall` header as soon as **the first delta carrying a non-empty `function.name`** arrives:

```ts
const buffered = bufferedByIndex.get(streamIndex) ?? { arguments: '', emitted: false };
...
if (!buffered.emitted) {
  if (!hasConcreteName) { /* buffer args */ return []; }
  buffered.emitted = true;                 // <-- committed on the FIRST name fragment
  ...
  return [toolCallHeader];                 // name = this fragment only
}
if (!hasArguments) return [];              // <-- any later name fragment is discarded here
```

The OpenAI-compatible streaming contract does not guarantee that `function.name` arrives in a single chunk, and it does not guarantee `name`-then-`arguments` ordering. When a provider splits the name, every fragment after the first is dropped, and arguments buffered before the name is complete are lost with it.

The same defect exists, unfixed, in the v1 copy at `packages/kosong/src/providers/chat-completions-stream.ts` (consumed by `packages/agent-core` and `packages/node-sdk`, i.e. the CLI path).

### Proposed fix

Accumulate the name; emit only once the `arguments` field appears (present, even empty — `typeof functionArguments === 'string'`, not `length > 0`, which is what distinguishes "field omitted" from "field present but empty" per `openai-python`'s `ChoiceDeltaToolCallFunction`); add a stream-end flush for calls that never receive an `arguments` field (vLLM omits it in some paths).

```ts
export interface BufferedChatCompletionToolCall {
  id?: string;
  name: string;          // added
  arguments: string;
  emitted: boolean;
}

if (!buffered.emitted) {
  if (hasConcreteName) buffered.name += functionName;
  const startsArguments = typeof functionArguments === 'string';
  if (buffered.name.length === 0 || !startsArguments) {
    if (hasArguments) buffered.arguments += functionArguments;
    bufferedByIndex.set(streamIndex, buffered);
    return [];
  }
  buffered.emitted = true;
  const initialArguments = buffered.arguments + functionArguments;
  buffered.arguments = '';
  bufferedByIndex.set(streamIndex, buffered);
  return [header(buffered, streamIndex, initialArguments)];
}
```

plus `flushChatCompletionStreamToolCalls(bufferedByIndex)` called after the `for await` loop in `_convertStreamResponse` (`openai-legacy.ts`), inside the `try`.

Verified locally: 6 unit tests covering split names, arguments-before-name, empty `arguments`, two concurrent indices, and end-of-stream flush. Three of them fail on the current code, all pass with the patch. Full `agent-core-v2` suite stays green (5775 tests).

### Two related silent-drop paths found while fixing this

- `kosong/contract/generate.ts`: a `tool_call_part` carrying **no** `index` is dropped without error when something (text, thinking) has been flushed between the header and the fragment — `flushPart` handles only `isContentPart` and `isToolCall`. Result: truncated argument JSON, no diagnostic.
- `agent/loop/loopService.ts`: `callsByIndex.set(part._streamIndex, …)` is called even when `_streamIndex` is `undefined`, so two index-less calls collide on the `undefined` key and later fragments are attributed to the wrong tool in the transcript.

---

### Symptom

The model states its intent in the thinking block and emits an unrelated trivial shell command instead of the tool it just described. Excerpt from a **fresh** CLI session (`kimi -p "…"`, small context), asked to write a file with `Write`:

| step | thinking | emitted call |
|---|---|---|
| 1 | *"The user asks for a documentation file, rapport.md…"* | `Bash mkdir -p … && cat > …` |
| 2 | *"**Now write the rapport.md in French, long and detailed.**"* | `Bash pwd` |
| 3 | *(empty)* | `Bash ls rapport.md` → `absent` |

The file is never created. `Write`, `Edit` and `Agent` are never emitted.

In a longer session the pattern dominates the run: 117 tool calls, of which 100 `Bash` — **54 with `command: "true"`**, 20 with `command: ""`, 5 with `command: "EOF"` — and zero `Write`/`Edit`/`Agent`. `"true"` and `""` are no-ops: the engine reports success, the model believes it acted, and the turn advances without any effect.

Two further observations:

- The thinking block collapses to **empty** on the degenerate steps, on a model that declares `always_thinking` with `effort = "high"`. 39 of 130 steps had empty thinking in one session.
- `finishReason` is a clean `tool_calls` / `stop` throughout — no truncation, no API error. The request is byte-identical across healthy and degenerate steps (same system-prompt hash, same tools hash, same params); only the message history grows.

### Note

vLLM's K3 deployment notes describe a matching model-side behaviour: *"We've occasionally seen K3 emit a tool-call format its own parser does not expect, yielding an empty `tool_calls` result"*, described as *"prompt- and run-dependent, not a blanket failure"*, with the recommendation to **validate against the schema and retry or fall back when `tool_calls` comes back empty**, or to use strict/structured tool calling.
<https://vllm.ai/blog/2026-07-27-k3>

The engine currently treats a degenerate call as a successful step. `loopControl.maxAttemptsPerStep` already exists but is never engaged for this case, because nothing classifies "empty or no-op tool call" as a failure.

### Suggested handling

1. Classify an empty / unparseable `tool_calls` payload as a retryable step failure so the existing attempt budget applies.
2. Optionally expose strict / structured tool calling for providers that support constrained decoding.
3. Surface a warning when a turn produces consecutive no-effect tool calls, instead of silently reporting success.

---

## Reproduction

```bash
cd "$(mktemp -d)"
kimi -p "Write a file report.md containing a full page of documentation about the architecture of an HTTP server in TypeScript: introduction, layer diagram, code samples, error handling section. Make it long and detailed. Use the Write tool."
ls report.md   # absent
```

Then inspect `~/.kimi-code/sessions/<workspace>/<session>/agents/main/wire.jsonl`:

```bash
jq -r 'select(.event.name) | .event.name' wire.jsonl | sort | uniq -c
jq -c 'select(.event.name=="Bash") | .event.args' wire.jsonl
```

Bug 1 reproduces more reliably with a large MCP catalog; Bug 2 appeared in a fresh session with a near-empty context, so it is not a context-length effect.

**Comment by sachnelson** [comment]
Independent reproduction, plus one new data point that extends the diagnosis.

**Environment:** kimi 0.36.0, Windows 11, provider `managed:kimi-code` (k3, thinking effort high), 8 MCP servers / ~130 tools. The Linear http server (55 tools, long shared `mcp__linear__` prefix) is the usual victim.

**Bug 1 confirmed in the wild.** Session wires from Aug 24 onward show the exact signature:

```json
{"type":"tool.call","name":"mcp__linear__list_issu","args":{}}
{"type":"tool.call","name":"mcp__linear__get_issu","args":{}}
```

First name fragment only, arguments dropped, `Tool "..." not found`, retry loop. Zero occurrences in months of session history before Aug 24, then 10+ per affected session. The client (0.36.0, released Aug 13) had been stable for ~10 days prior, and this issue was opened the same day the failures started here — pointing at a serving-side streaming change around Aug 23–24 rather than a client regression.

**New evidence: the name fragments themselves are malformed.** I applied the proposed patch locally (both copies — `packages/kosong` and `packages/agent-core-v2` — on main @ 41f1eae). The parser fix works as designed, and it exposed what the stream actually contains. The same call now records the fully reassembled name as:

```json
{"type":"tool.call","name":"mcp__linear__list_issu?list_issues","args":{}}
```

Hex dump confirms a literal `?` (0x3F) between the fragments — the continuation delta carries `?list_issues`, not `es`. So the old parser wasn't truncating a valid name; it was amputating a name that arrives corrupt from the server. This is Bug 2 / the vLLM K3 emission note surfacing through the Bug 1 code path. Telling detail: the model's own text read "Typo in tool name. Use mcp__linear__list_issues." — and it then emitted the identical corrupted name again. It knows the right name; the serving layer can't serialize it.

**Patch verification (Windows, Node 25.5.0):** applied the reporter's approach to both engine copies, plus a `flushChatCompletionStreamToolCalls` at the three call sites (`kimi.ts`, `openai-legacy.ts` in both packages). 8 new unit tests per package (split name, args-before-name, empty `arguments: ""`, interleaved parallel indices, no-index passthrough, flush semantics) — all pass; full kosong suite 1373/1373; agent-core-v2 kosong tests pass; `tsc --noEmit` and oxlint clean. (For the record: the full agent-core-v2 suite shows 153 failures on Windows in fs/symlink/path domains — verified byte-identical against a pristine clone, so pre-existing platform issues, unrelated to this change.)

Happy to open this as a PR with the tests once the issue gets an `/approve`, per the contributing guide. Filed #3230 for the server-side corruption + a client-side validation proposal, since that part is out of scope for the parser fix.

**Comment by houleixx2** [comment]
The new `?list_issues` observation makes raw delta preservation especially important. For one affected call, capture every `tool_calls[index]` fragment with byte-exact `id`, `function.name`, and `arguments`, then compare three values: the concatenation implied by the wire, the engine's assembled call, and the declared tool name from the request. That separates a client-side early-commit bug from a provider-side corrupt continuation, and it gives the validator a precise reason to retry rather than executing a near-match or no-op call.

I maintain ccglass, an open-source OpenAI-compatible request/response inspector. It can retain the tool schema and streamed response in one redacted trace and export the raw exchange for parser fixtures. It cannot observe model logits or correct a malformed server fragment, but it can establish which bytes crossed the API boundary and whether the engine changed them afterward.

Project: https://github.com/jianshuo/ccglass

**Comment by fusengine** [comment]
## New evidence: the defect is present from turn 1, on an empty context

I went back to the wire log with a full inventory of emitted tool names rather than
spot checks, and the result contradicts the context-degradation explanation used in
#750 and #3214. Posting it here because it narrows the search considerably.

### Full inventory — one session, 124 tool calls

| Tool name emitted | Count |
| --- | --- |
| `Bash` | 87 |
| `mcp__fuse-browser__browser_open` | 24 |
| `Read` | 5 |
| `FetchURL` | 3 |
| `mcp__exa__web_search_exa` | 2 |
| `mcp__fuse-browser__browser_opem` *(corrupted)* | 2 |
| `Grep` | 1 |
| **`Write` / `Edit` / `Agent`** | **0** |

34 of the 87 `Bash` calls carry `{"command": "true"}` — a shell no-op that always
returns "Command executed successfully", so nothing surfaces as a failure and the
model retries indefinitely.

### The part that matters

- first `Bash {"command":"true"}` — **turn 1**
- last — turn 34
- first corrupted tool name (`browser_opem`) — **turn 1**

The very first tool call of the session, on an almost empty context, is already
wrong. There is no accumulation and no threshold to cross. Whatever this is, it is
not progressive context degradation.

Two further details that constrain the cause:

1. `browser_opem` vs `browser_open` is a **substituted character**, not a truncation.
   A dropped SSE delta cannot produce it — the name is the right length with one
   wrong character. So this is not a streaming-assembly artifact (I had assumed it
   was, and patched the assembler accordingly; the patch is correct on its own merits
   but does not address this).
2. `command: "true"` is a suspiciously well-chosen no-op. It reads like a fallback
   value rather than a hallucinated command.

### Ruled out on this session

- **The tool is sent.** `Write` is present in every request with its full description
  ("Create, append to, or replace a file entirely"), verified in the actual outgoing
  payload.
- **No pruning.** `toolSelect: false` on all 152 requests — the complete tool list
  goes out every time.
- **No client-side interference.** No hook blocked anything, no tool returned an
  error, no permission was denied.

### Environment

Kimi Code CLI 0.38.0 · `kimi-code/k3`, thinking effort `high`, max context 1048576 ·
macOS Darwin 25.6.0 arm64 · Node 26.7.0 · ~155 tools exposed, ~248 KB of tool schemas.

### Question

The vLLM deployment notes for K3 mention the model occasionally emitting a tool-call
format its own parser does not expect, yielding an empty `tool_calls` result. Is
`Bash {"command":"true"}` what that fallback materialises as downstream? If so, the
fix belongs on the inference side, and clients would at minimum need a way to detect
a degenerate call instead of executing a silent success.

Full wire log available on request.

**Comment by Tolik-Trek** [comment]
**Bug 2 reproduced in the wild — two independent sessions, Windows 11, CLI 0.39.1, model `kimi-code/k3` (provider `managed:kimi-code`).**

**Session A** (interactive TUI, swarm mode, thinking effort `max`): the main agent intended a single `AgentSwarm` call (long `prompt_template` + 8 items). Steps 27–42 of the turn: thinking blocks read "NOW. AgentSwarm. Single tool call. Go", "I keep stalling with pointless bash calls", "I appear to be stuck in a pathological loop of emitting trivial Bash calls", "Something is seriously wrong — I've emitted about 8 trivial bash calls in a row" — while the emitted calls were `Bash {"command":"true"}`, `echo launching-wave-A`, `echo ok`, `echo done`, `echo final-check-before-swarm`, `echo swarm-A-launch`, `echo AGENTSWARM-NEXT`, `echo LAST-ONE`. 15+ consecutive degenerate steps; the user cancelled the turn after ~17 minutes. Zero `AgentSwarm`/`Agent` calls ever reached the harness — no tool errors anywhere in the wire; every no-op returned `Command executed successfully`.

**Session B** (goal mode, context ~207k tokens cached input): 8 consecutive `Bash {"command":"true"}` instead of an intended `Agent` launch. The harness anti-repetition reminders escalated ("repeated several times" → 5th → 6th → 7th → forced final response) without breaking the loop. Thinking blocks on degenerate steps were **empty** (matches the report), `finishReason: tool_use` throughout. Earlier in the same session, background `Agent` launches worked fine (two subagents completed) — the tool itself was functional.

**Model self-diagnosis** (asked afterwards in session B): "баг был на моей стороне, не в харнесе… я собирался запустить субагента с очень длинным промптом (~100 строк JSON)… вместо того чтобы сгенерировать этот большой payload, я эмитировал самый дешёвый вызов из возможных: `true`… После первого `true` в контексте появилась пара «запускаю субагента → Bash true», и каждый следующий ход воспроизводил этот паттерн." I.e. a repetition loop seeded by one degenerate emission, exactly the mechanism described above.

**Two additional data points:**

1. Prompt-level mitigations do NOT reliably break the loop. Anti-loop rules were injected into the session via `AGENTS.md` ("intention = immediate tool call, first in the response", "long brief → write to file first", "a no-op call is a red flag — stop") — the loop recurred later in the same session anyway. The only reliable break was the user interrupting the turn (Esc).
2. Live meta-reproduction: while preparing THIS comment, the main agent (k3-256k, CLI 0.39.1) intended `add_issue_comment` on this issue and instead emitted `list_issue_fields` **7 times in a row**, then two wrong `issue_write` calls (creating two accidental issues, since closed), escalating through the same harness reminders. It only broke the pattern by switching channels entirely (`gh` CLI instead of the MCP tool). The "intended expensive/specific call → degenerate repeated nearby call" failure is therefore not limited to `Agent`/`Write` — it also hits MCP tool selection.

Workaround that has held so far: keep the intended tool call small (brief-to-file + short prompt referencing the path), avoid one giant `AgentSwarm` template in favor of several small parallel `Agent` calls, and on the first no-op emission treat it as a red flag — interrupt rather than retry, because the pattern self-reinforces from context.

Session wires (both sessions + the meta-reproduction session) available on request.
